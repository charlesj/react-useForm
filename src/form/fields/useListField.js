import { List, Map } from 'immutable'
import { actions } from '../reducer/fieldReducer'
import { resolveFieldData, getFieldProps } from '../resolveFieldData'
import { getFields } from '../reducer/generateDefaultListState'
/**
 * @param  fieldArgs { import("../useForm").IFormFieldArgs}
 * @returns {import("../useForm").IFormField}
 */
export const useListField = (state, dispatch, fieldArgs = {}) => {
  const requiredMessage = fieldArgs.requiredMessage || 'Required'

  const setValidationResult = result => {
    dispatch(actions.validationResult(fieldArgs.name, true, result))
  }

  const setValue = v => {
    // dispatch(actions.updateValue(fieldArgs.name, v))
    // tryValidate(v, state.getIn(['current', 'touched']))
  }

  const validate = () => {
    const isEmpty = state.get('items', List()).size === 0
    dispatch(actions.validationResult(fieldArgs.name, isEmpty, isEmpty ? requiredMessage : ''))
    return !isEmpty
  }

  const add = (item = Map()) => {
    const options = state.getIn(['initial', 'options'])
    const field = state.getIn(['initial', 'field'])
    const size = state.get('items', List()).size
    const itemState = getFields(field.fields, item, options, fieldArgs.name, size)
    dispatch(actions.addListItem(fieldArgs.name, itemState))
  }

  const remove = index => {
    dispatch(actions.removeListItem(fieldArgs.name, index))
  }

  const fieldData = state.getIn(['items']).map(item => {
    return resolveFieldData(item, dispatch)
  }).toJS()

  return {
    props: {
      error: state.getIn(['current', 'error']),
      helperText: state.getIn(['current', 'helperText']),
      label: state.getIn(['initial', 'label']),
      items: fieldData.map((item, index) => ({
        ...getFieldProps(item),
        key: `${fieldArgs.name}.${index}`,
      })),
      add,
      remove,
    },
    setValidationResult,
    setValue,
    validate,
  }
}

export const defaultListValue = List()
