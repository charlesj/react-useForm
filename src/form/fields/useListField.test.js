import { fromJS } from 'immutable'
import { useListField } from './useListField'

const dispatch = jest.fn()
let state, fieldArgs
beforeEach(() => {
  fieldArgs = {
    name: 'fieldName',
    type: 'list',
    fields: [
      { name: 'first', label: 'First' },
    ],
  }

  state = fromJS({
    initial: {
      label: 'the label',
      type: 'list',
    },
    current: {
      error: false,
      helperText: 'help me',
    },
    items: [
      {
        fields: {
          first: {
            initial: {
              type: 'text',
              label: 'First',
              field: 'initialFirstField',
            },
            current: {
              error: 'false',
              helperText: '',
              value: '1',
            },
          },
          second: {
            initial: {
              type: 'text',
              label: 'Second',
              field: 'initialSecondField',
            },
            current: {
              error: 'false',
              helperText: '',
              value: '2',
            },
          },
        },
      },
    ],
  }).setIn(['initial', 'options'], {})
    .setIn(['initial', 'field'], fieldArgs)

})

test('returns props needed', () => {
  expect(useListField(state, dispatch, fieldArgs)).toMatchSnapshot()
})

test('setValidationResult externally', () => {
  const { setValidationResult } = useListField(state, dispatch, fieldArgs)

  setValidationResult('i am error')
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('add', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.add()
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('add with values', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.add(fromJS({ first: '111' }))
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})

test('remove', () => {
  const { props } = useListField(state, dispatch, fieldArgs)
  props.remove(1)
  expect(dispatch.mock.calls[0]).toMatchSnapshot()
})


describe('validate', () => {
  test('has items', () => {
    const { validate } = useListField(state, dispatch, fieldArgs)
    expect(validate()).toBeTruthy()
    expect(dispatch.mock.calls[0]).toMatchSnapshot()
  })

  test('no items', () => {
    const nextState = state.set('items', fromJS([]))
    const { validate } = useListField(nextState, dispatch, fieldArgs)
    expect(validate()).toBeFalsy()
    expect(dispatch.mock.calls[0]).toMatchSnapshot()
  })
})
