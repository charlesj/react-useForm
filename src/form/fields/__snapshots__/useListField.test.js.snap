// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Immutable.Map {
      "fields": Immutable.Map {
        "first": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "value": "",
            "optional": false,
            "label": "First",
            "field": Object {
              "label": "First",
              "name": "fieldName.items.1.fields.first",
            },
          },
          "current": Immutable.Map {
            "helperText": "",
            "error": false,
            "pristine": true,
            "touched": false,
            "value": "",
          },
        },
      },
    },
    "type": "addListItem",
  },
]
`;

exports[`add with values 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Immutable.Map {
      "fields": Immutable.Map {
        "first": Immutable.Map {
          "initial": Immutable.Map {
            "type": "text",
            "value": "111",
            "optional": false,
            "label": "First",
            "field": Object {
              "label": "First",
              "name": "fieldName.items.1.fields.first",
            },
          },
          "current": Immutable.Map {
            "helperText": "",
            "error": false,
            "pristine": true,
            "touched": false,
            "value": "111",
          },
        },
      },
    },
    "type": "addListItem",
  },
]
`;

exports[`remove 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": 1,
    "type": "removeListItem",
  },
]
`;

exports[`returns props needed 1`] = `
Object {
  "props": Object {
    "add": [Function],
    "error": false,
    "helperText": "help me",
    "items": Array [
      Object {
        "first": Object {
          "error": "false",
          "helperText": "",
          "label": "First",
          "onBlur": [Function],
          "onChange": [Function],
          "value": "1",
        },
        "key": "fieldName.0",
        "second": Object {
          "error": "false",
          "helperText": "",
          "label": "Second",
          "onBlur": [Function],
          "onChange": [Function],
          "value": "2",
        },
      },
    ],
    "label": "the label",
    "remove": [Function],
  },
  "setValidationResult": [Function],
  "setValue": [Function],
  "validate": [Function],
}
`;

exports[`setValidationResult externally 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": true,
      "helperText": "i am error",
    },
    "type": "validationResult",
  },
]
`;

exports[`validate has items 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": false,
      "helperText": "",
    },
    "type": "validationResult",
  },
]
`;

exports[`validate no items 1`] = `
Array [
  Object {
    "fieldName": "fieldName",
    "payload": Object {
      "error": true,
      "helperText": "Required",
    },
    "type": "validationResult",
  },
]
`;
